onnxscript/ir/passes/common/common_subexpression_elimination.py

```diff
@@ -0,0 +1,163 @@
+            doc_string=old_graph.doc_string,
+        )
+        # 2. Create a mapping from olds to news.
+        old_node_hash_to_new_node: dict[int, ir.Node] = {}
+        old_value_to_new_value: dict[ir.Value, ir.Value] = {}
+
+        # 3. Create inputs and initializers in the new graph
+        #    from the old graph.
+        for input in old_graph.inputs:
+            new_input = _copy_value(input)
+            new_graph.inputs.append(new_input)
+            old_value_to_new_value[input] = new_input
+            logger.debug("Adding input %s", new_input.name)
+        for initializer in old_graph.initializers.values():
+            new_initializer = _copy_value(initializer)
+            new_graph.register_initializer(new_initializer)
+            old_value_to_new_value[initializer] = new_initializer
+            logger.debug("Adding initializer %s", new_initializer.name)
+
+        # 4. Create nodes in the new graph from the old graph.
+        for old_node in old_graph:
+            # 4.0. Iterate and update node inputs.
+            old_node_inputs: list[ir.Value] = []
+            for old_input in old_node.inputs:
+                assert old_input is not None
+                if old_input in old_value_to_new_value:
+                    old_node_inputs.append(old_value_to_new_value[old_input])
+                else:
+                    old_node_inputs.append(old_input)
+            # 4.1. Construct the (node, inputs, attributes) hash to
+            #      check if the node is a common subexpression.
+            # Attr is not hashable
+            attributes = {}
+            for k, v in old_node.attributes.items():
+                assert isinstance(v, ir.Attr)
+                attributes[k] = v.value
+            hash_value = hash(
+                (
+                    old_node.op_identifier(),
+                    tuple(old_node_inputs),
+                    tuple(sorted(attributes.items())),
+                )
+            )
+            # TODO(titaiwang): Subgraphs are not supported yet.
+            # TODO(titaiwang): Skip control flow nodes?
+            # 4.2. Check if the node is a common subexpression.
+            if hash_value in old_node_hash_to_new_node:
+                # 4.2.1. If it is, this node is already in the new graph, so
+                #         we don't need to create a new node.
+                modified = True
+                new_node = old_node_hash_to_new_node[hash_value]
+                logger.debug("Reusing node %s", new_node.name)
+            else:
+                # 4.2.2. If it is not, create a new node and add it to the graph.
+                new_node = _copy_node(old_node, old_value_to_new_value)
+                new_graph.append(new_node)
+                old_node_hash_to_new_node[hash_value] = new_node
+            # 4.3 Add the node outputs to the mapping.
+            old_value_to_new_value.update(dict(zip(old_node.outputs, new_node.outputs)))
```

Comment:
We will need to handle the possibility of optional outputs as well: that is a minor complication. We need to encode the outputs-present as an extra entry in the tuple used above. If we have two nodes which look identical but one has an optional output and the other doesn't, the above code could break. So, better to not treat them as being same. So, `output1 = SomeOp(A, B)` and `outputX, outputY = SomeOp(A, B)` will be treated as being different. Unlikely to happen, but good to handle it.

Reply:
Do you have an example op in mind? I am not sure if the following is the optional input case you have in mind:

```python
        def f(x)
            a, b = topk(x, k=1)
            c, _ = topk(x, k=1)
            return a, b, c
```

Reply:
I can't think of an op that could have different number of outputs when the inputs and attributes are the same. Could you think of any?

Reply:
I added len(node.outputs) to op_info (as unique identifier) but do not think of a test to guard it. however, it should work.